hack/collect-logs.sh

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env bash
+
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+CLUSTER_NAME=${CLUSTER_NAME:-"test-$(date +%s)"}
+ARTIFACTS="${ARTIFACTS:-${PWD}/_artifacts}"
+
+# dump logs from kind and all the nodes
+dump-logs() {
+  # log version information
+  echo "=== versions ==="
+  echo "kind : $(kind version)" || true
+  echo "kubectl: "
+  kubectl --kubeconfig=$CLUSTER_NAME.kubeconfig version || true
+  echo ""
+
+  # dump all the info from the CAPI related CRDs
+  mkdir -p $ARTIFACTS/logs
+  kubectl get \
+  clusters,azureclusters,machines,azuremachines,kubeadmconfigs,machinedeployments,azuremachinetemplates,kubeadmconfigtemplates,machinesets,kubeadmcontrolplanes \
+  --all-namespaces -o yaml >> "${ARTIFACTS}/logs/capz.info" || true
+
+  # dump images info
+  {
+   echo "images in docker"
+   docker images
+   echo "images from bootstrap using containerd CLI"
+   docker exec kind-control-plane ctr -n k8s.io images list
+   echo "images in bootstrap cluster using kubectl CLI"
+   (kubectl get pods --all-namespaces -o json \
+   | jq --raw-output '.items[].spec.containers[].image' | sort)
+   echo "images in deployed cluster using kubectl CLI"
+   (kubectl --kubeconfig=$CLUSTER_NAME.kubeconfig get pods --all-namespaces -o json \
+   | jq --raw-output '.items[].spec.containers[].image' | sort)
+ } >> "${ARTIFACTS}/logs/images.info"
+  
+  # dump cluster info for kind
+  {
+    echo "kind cluster-info"
+    kubectl cluster-info dump
+  } >> "${ARTIFACTS}/logs/kind-cluster.info"
+
+  # dump cluster info for capz
+  {
+    echo "=== VMs in ${AZURE_RESOURCE_GROUP}  ==="
+    az vm list --resource-group "${AZURE_RESOURCE_GROUP}"
+    echo "=== cluster-info dump ==="
+    kubectl --kubeconfig=$CLUSTER_NAME.kubeconfig cluster-info dump
+  } >> "${ARTIFACTS}/logs/capz-cluster.info"
+
+  # export all logs from kind
+  kind "export" logs --name="kind" "${ARTIFACTS}/logs"
+
+  nodes=$(az vm list --resource-group ${AZURE_RESOURCE_GROUP} --query "[?tags.\"sigs.k8s.io_cluster-api-provider-azure_cluster_capi-quickstart\" == 'owned'].name" -o tsv)
```

Comment:
Nice query!

Reply:
Thank you! This query inspired me to go down the zsh rabbit hole for syntax highlighting :)